lib/charms/alertmanager_k8s/v0/alertmanager_remote_configurer.py

```diff
@@ -0,0 +1,1888 @@
+    Separation of the main configuration and the templates is dictated by the assumption that
+    the default provider of the `alertmanager_remote_configurer` relation will be
+    `alertmanager-k8s` charm, which requires such separation.
+    """
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        relation_name: str = DEFAULT_RELATION_NAME,
+        config_file_path: str = DEFAULT_ALERTMANAGER_CONFIG_FILE_PATH,
+    ):
+        """API that manages a required `remote-configurer` relation.
+
+        Args:
+            charm: The charm object that instantiated this class.
+            relation_name: Name of the relation with the `alertmanager_remote_configurer` interface
+                as defined in metadata.yaml. Defaults to `remote-configurer`.
+            config_file_path: The path to the Alertmanager configuration file. Defaults
+                to `/etc/alertmanager/alertmanager.yml`.
+        """
+        super().__init__(charm, relation_name)
+        self._charm = charm
+        self._relation_name = relation_name
+        self._config_file_path = config_file_path
+
+        on_relation = self._charm.on[self._relation_name]
+
+        self.framework.observe(on_relation.relation_joined, self._on_relation_event)
+        self.framework.observe(on_relation.relation_changed, self._on_relation_event)
+        self.framework.observe(self._charm.on.upgrade_charm, self._on_upgrade_charm)
+
+    def _on_relation_event(self, event: Union[RelationJoinedEvent, RelationChangedEvent]) -> None:
+        """Event handler for remote configurer's relation events.
+
+        Takes care of pushing Alertmanager configuration to the relation data bag.
+
+        Args:
+            event: Juju event
+
+        Returns:
+             None
+        """
+        if not self._charm.unit.is_leader():
+            return
+        self._update_relation_databag(event.relation)
+
+    def _on_upgrade_charm(self, _) -> None:
+        """Event handler for charm upgrade event.
+
+        Takes care of pushing Alertmanager configuration to the relation data bag.
+
+        Returns:
+             None
+        """
+        if not self._charm.unit.is_leader():
+            return
+        relation = self.model.get_relation(self._relation_name)
+        self._update_relation_databag(relation)
+
+    def _update_relation_databag(self, relation: Optional[Relation]) -> None:
+        try:
+            self._config = load_config_file(self._config_file_path)
+            self._templates = self._get_templates(self._config)
+            relation.data[self._charm.app]["alertmanager_config"] = json.dumps(self._config)  # type: ignore[union-attr]  # noqa: E501
+            relation.data[self._charm.app]["alertmanager_templates"] = json.dumps(self._templates)  # type: ignore[union-attr]  # noqa: E501
+        except FileNotFoundError as e:
+            self._charm.unit.status = BlockedStatus(str(e))
```

Comment:
Libraries should not block. Emit an error event the charm author can chose to subscribe to and block if they think it's necessary.

Reply:
Removing reading from disk in the constructor actually got me that for free :)
Done.